core_lib/src/interface/scribblearea.cpp

```diff
@@ -573,28 +586,35 @@ void ScribbleArea::handleDoubleClick()
 
 bool ScribbleArea::isLayerPaintable() const
 {
-    if (!areLayersSane())
+    if (!isKeyframeSane())
         return false;
 
     Layer* layer = mEditor->layers()->currentLayer();
     return layer->type() == Layer::BITMAP || layer->type() == Layer::VECTOR;
 }
 
-bool ScribbleArea::areLayersSane() const
+bool ScribbleArea::isKeyframeSane() const
 {
     Layer* layer = mEditor->layers()->currentLayer();
     // ---- checks ------
     if (layer == nullptr) { return false; }
-    if (layer->type() == Layer::VECTOR)
-    {
-        VectorImage* vectorImage = currentVectorImage(layer);
-        if (vectorImage == nullptr) { return false; }
+
+    int action = mPrefs->getInt(SETTING::DRAW_ON_EMPTY_FRAME_ACTION);
+    KeyFrame* lastKeyframe = layer->getLastKeyFrameAtPosition(mEditor->currentFrame());
+    KeyFrame* currentKeyframe = layer->getKeyFrameAt(mEditor->currentFrame());
+
+    // the first frame has to exist to use this action
+    if (action == DUPLICATE_PREVIOUS_KEY) {
+        if (lastKeyframe == nullptr) { return false; }
     }
-    if (layer->type() == Layer::BITMAP)
-    {
-        BitmapImage* bitmapImage = currentBitmapImage(layer);
-        if (bitmapImage == nullptr) { return false; }
+    if (currentKeyframe == nullptr) {
+        // We can assume keyframe is sane here because a
+        // keyframe will be created before drawing on the canvas.
+        if (action != (CREATE_NEW_KEY | DUPLICATE_PREVIOUS_KEY)) {
+            return false;
+        }
     }
```

Comment:
I'm confused about what you are trying to do here. A keyframe will be created if `action == CREATE_NEW_KEY`. I think it's also plausible that `DUPLICATE_PREVIOUS_KEY` should act like create new key if there is no previous key (duplicating the currently shown empty image) but you already return false for that above. When `action == KEEP_DRAWING_ON_PREVIOUS_KEY`, then there's a problem and it should return false. If I was doing this I would just use the last condition I just mentioned. Anyway, whatever you want to do, it doesn't make sense to use `|` because `CREATE_NEW_KEY` and `DUPLICATE_PREVIOUS_KEY` are not set to be bit flags. CREATE_NEW_KEY is 0 and DUPLICATE_PREVIOUS_KEY is 1, so this currently just evaluates to `action != DUPLICATE_PREVIOUS_KEY`. This causes all input to be ignored when using the "Create new key" behavior.

Reply:
Yeah that's completely bogus... I guess that's what I get from trying to program, late at night...

The idea was basically that if we're on the last keyframe, then we check for if it was DUPLICATE_PREVIOUS otherwise we look at the current frame and check if it's CREATE_NEW_KEY or DUPLICATE_PREVIOUS_KEY. The bitwise condition is bogus, not sure what I thought there.

I'll rework it and yeah it makes much more sense to look at the last action instead.